driver/src/price_estimation/mod.rs

```diff
@@ -0,0 +1,456 @@
+//! Module responsible for aggregating price estimates from various sources to
+//! give good price estimates to the solver for better results.
+
+#![allow(dead_code)]
+
+mod kraken;
+
+use self::kraken::KrakenClient;
+use crate::contracts::erc20::ERC20Detailed;
+use crate::contracts::stablex_contract::{BatchExchange, StableXContract};
+use crate::util::FutureWaitExt;
+use anyhow::Result;
+use ethcontract::Address;
+use lazy_static::lazy_static;
+use log::warn;
+use serde::{Serialize, Serializer};
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::iter::FromIterator;
+
+/// A price estimate result for the given tokens.
+///
+/// This type implements JSON deserialization so that it is accepted as input to
+/// the solver.
+#[cfg_attr(test, derive(Eq, PartialEq))]
+#[derive(Clone, Debug, Default, Serialize)]
+#[serde(transparent)]
+pub struct PriceEstimates(pub BTreeMap<TokenId, u128>);
+
+impl PriceEstimates {
+    /// Create a new price estimate mapping from an iterator of estimates.
+    fn new<I>(estimates: I) -> Self
+    where
+        I: IntoIterator<Item = (TokenId, u128)>,
+    {
+        PriceEstimates(BTreeMap::from_iter(estimates))
+    }
+}
+
+/// A price oracle to retrieve price estimates for exchange tokens to help the
+/// solver find better solutions.
+pub struct PriceOracle {
+    /// The exchange contract.
+    contract: Box<dyn StableXContract>,
+    /// The token reader for retrieving token information.
+    token_reader: Box<dyn TokenReader>,
+    /// The price source being used.
+    price_source: Box<dyn PriceSource>,
+
+    /// Cache of token information that is used for
+    tokens: Vec<Token>,
+    /// Cached count of read tokens.
+    read_token_count: u16,
+}
+
+lazy_static! {
+    /// Tokens that are pegged to USD.
+    static ref USD_TOKENS: HashSet<String> = HashSet::from_iter(
+        ["USDT", "TUSD", "USDC", "PAX", "GUSD", "DAI", "sUSD"]
```

Comment:
These are just used as a fallback price in case Kraken does not list prices for them. "PAX" for example is not listed on Kraken so it gets the default `10**18` value.

Note that I tried to set this up in a way that the price oracle can have multiple `PriceSource`s, where Kraken is one of them. So hopefully, as more price sources are included, the fallback prices will not have to be used.